PayPalWebPayments/src/test/java/com/paypal/android/paypalwebpayments/PayPalWebCheckoutClientUnitTest.kt

```diff
@@ -7,53 +7,80 @@ import com.braintreepayments.api.BrowserSwitchClient
-import io.mockk.mockk
-import io.mockk.slot
-import io.mockk.verify
+import io.mockk.*
 import junit.framework.TestCase.assertEquals
 import junit.framework.TestCase.assertNull
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.test.advanceUntilIdle
 import org.json.JSONObject
 import org.junit.Test
 import org.junit.runner.RunWith
 import org.robolectric.RobolectricTestRunner
+import strikt.api.expectThat
+import strikt.assertions.isEqualTo
 import java.lang.reflect.Field
+import kotlinx.coroutines.test.*
 
+@ExperimentalCoroutinesApi
 @RunWith(RobolectricTestRunner::class)
 class PayPalWebCheckoutClientUnitTest {
 
     private val browserSwitchClient: BrowserSwitchClient = mockk(relaxed = true)
     private val browserSwitchHelper: BrowserSwitchHelper = mockk(relaxed = true)
     private val activity: FragmentActivity = mockk(relaxed = true)
     private val coreConfig: CoreConfig = mockk(relaxed = true)
+    private val api = mockk<API>(relaxed = true)
 
     @Test
-    fun `approveOrder starts browserSwitchClient with correct parameters`() {
-        val payPalClient =
-            PayPalWebCheckoutClient(activity, coreConfig, browserSwitchClient, browserSwitchHelper)
-        val payPalRequest = PayPalWebCheckoutRequest("mock_order_id")
-        val browserSwitchOptions = mockk<BrowserSwitchOptions>(relaxed = true)
+    fun `start() delivers error if error fetching clientID`() = runTest {
+        val fakeCode = 123
+        val error = PayPalSDKError(fakeCode, "fake-description")
+        val errorSlot = slot<PayPalSDKError>()
+        val payPalCheckoutListener = spyk<PayPalWebCheckoutListener>()
 
+        coEvery { api.fetchCachedOrRemoteClientID() } throws error
         every {
-            browserSwitchHelper.configurePayPalBrowserSwitchOptions(
-                any(),
-                coreConfig,
-                payPalRequest.fundingSource
-            )
+            payPalCheckoutListener.onPayPalWebFailure(capture(errorSlot))
+        } answers { errorSlot.captured }
+
+        val sut = getPayPalCheckoutClient(testScheduler = testScheduler)
+        sut.listener = payPalCheckoutListener
+
+        sut.start(mockk(relaxed = true))
+        advanceUntilIdle()
+
+        verify {
+            payPalCheckoutListener.onPayPalWebFailure(any())
+        }
+        expectThat(errorSlot.captured) {
+            get { code }.isEqualTo(fakeCode)
+            get { errorDescription }.isEqualTo("Error fetching clientID. Contact developer.paypal.com/support.")
+        }
+    }
+
+    @Test
+    fun `start() starts browserSwitchClient with correct parameters`() = runTest {
+        val sut = getPayPalCheckoutClient(testScheduler = testScheduler)
+        val browserSwitchOptions = mockk<BrowserSwitchOptions>(relaxed = true)
+
+        coEvery { api.fetchCachedOrRemoteClientID() } returns "fake-client-id"
+
+        coEvery {
+            browserSwitchHelper.configurePayPalBrowserSwitchOptions(any(), any(), any())
```

Comment:
I'm not sure why, but the only way to get this test passing was to change from `any(), coreConfig, payPalRequest.fundingSource` params to `any(), any(), any()` for the stub 🤔 - none of the logic around browser switch options change.

Reply:
It's probably because of a [Referential Equality](https://kotlinlang.org/docs/equality.html) check in this case since the custom types don't have any `equals(other: Any?)` methods implemented.